### DISCUSSION TITLE
Test package for ecosystem libraries

### DISCUSSION THREAD

**Discussion by Fryuni** [withastro/roadmap#1021] [Proposal]
### Body

# Summary

This proposal suggests a new `@astrojs/test-utils` package (name TBD) containing first-party tools for library, adapter and integration authors to test their packages reliably. Such a package would enable authors to maintain their own projects without breaking package isolation nor having to replicate parts of Astro implementation on their own test suite.

# Background & Motivation

There are docs around [testing an Astro site](https://docs.astro.build/en/guides/testing/) but, currently, Astro lacks official tooling for testing integrations and libraries. There has been questions regarding this on GH and on Discord, for example:

- https://github.com/withastro/roadmap/discussions/74
- https://discord.com/channels/830184174198718474/1178906267989843998/1178906267989843998
- https://discord.com/channels/830184174198718474/1167265554101436488/1167265554101436488

Sometimes we point the users to the adapters repo as a source for them to copy into their own projects, but even within the adapters repo there [has been problems](https://discord.com/channels/830184174198718474/845430950191038464/1248544598570045491).

This lack of test tooling also affects official Astro projects outside of the `astro` package. [Issues](https://github.com/withastro/astro/issues/8960) and [discussions](https://discord.com/channels/830184174198718474/830184175176122389/1027614344856481942) about providing a test package having beeing going on for a while.

# Goals

- Provide an official testing utility specifically designed for Astro integrations, libraries, and components.
- Simplify the process of creating, running, and maintaining tests for projects on the Astro ecosystem.
- Not force any specific testing library (should be usable with vitest, Node tests, Playwright, Cypress, etc.)

# Examples

## Astro Core tests

`astro` has a `test-utils` module that defines a great set of utilities for testing.

A fixture is loaded and a collection of method to work with that fixture are provided. The API usability follows this style:

```ts
const fixture = await loadFixture({
	root: './fixtures/astro-basic/',
});
await fixture.build();
const previewServer = await fixture.preview();
```

This is a good starting point, but has some problems:

- It is focused on testing Astro internals, not libraries made to be used _with_ Astro.
- It relies heavily on Astro internals, which are not exposed for other packages to use.
- Astro's core tests are in JS, so there was no need to comply with TS for this test utils module. For arbitrary projects in the ecosystem, the fixture should be typed.

## Third party implementation

I ported the utilities from core into an external package to use for my own tests, [`@inox-tools/astro-tests`](https://inox-tools.fryuni.dev/astro-tests).

[Here](https://github.com/Fryuni/inox-tools/blob/c4176513b68066b33874e15345f439fae97b2e4e/packages/astro-when/tests/server-output.test.ts) is an example of it being used.

It mostly resolves the problem with the module from `astro`'s tests, but it still has some disadvantages that can be resolved from Astro core:

- The function `mergeConfig` that combines inline configuration with configuration loaded from an Astro config file is not exported. This function is similar to a function of the same name from Vite, but Vite's function _is_ exported. Ideally this function could be exported from Astro so it would not have to be copied into the package.
- Similarly, the function `validateConfig` that fills default values and apply transformations on the configuration is also not exported. This is required to resolve a value from the configuration that is not directly set on the test but used for the test tolling, like the `outDir` and `srcDir`.
- The exported `build` function hides the option to re-use the instantiated WASM module from the compiler. This causes a performance degradation on tests building projects.

**Comment by matthewp** [comment]
Whether we create a new package or not, we still need to expose these internals. Can you provided the suggested signatures and exports of what you'd like to see exposed? For example, `mergeConfig` is currently typed as `mergeConfig(defaults: Record<string, any>, overrides: Record<string, any>): Record<string, any>)`, is that ok? Also, where should this be exposed from? `astro/config`? Same question for the other internal functions.

**Reply by Fryuni** [reply]
I think both `mergeConfig` and `validateConfig` would fit nicely being exported from `astro/config`.

The signature of `mergeConfig` can be changed to this:
```ts
function mergeConfig<C extends AstroConfig | AstroInlineConfig>(
	defaults: C,
	overrides: DeepPartial<C>,
	isRoot = true,
): C
```

I didn't find any use in the codebase where that `isRoot` is passed, so it is effectively always true. We could remove it.

`validateConfig` already has the proper signature IMO, it accepts anything as the input and then validates and transforms the input, returning an `AstroConfig`:
```ts
function validateConfig(
	userConfig: any,
	root: string,
	cmd: string,
): Promise<AstroConfig>
```

Lastly, I don't think [this wrapper](https://github.com/withastro/astro/blob/1c64ae304d3c008a776e98e48fd3ece8be7b1fb5/packages/astro/src/core/index.ts?plain=1#L10-L17) to hide the second parameter is needed. Neither [this](https://github.com/withastro/astro/blob/1c64ae304d3c008a776e98e48fd3ece8be7b1fb5/packages/astro/src/core/build/index.ts#L43) internal modifier.

**Reply by matthewp** [reply]
Why `AstroConfig | AstroInlineConfig`? Is this running the schema to convert it?

**Reply by Fryuni** [reply]
It is used to merge values both before and after resolving.

- It merges `AstroInlineConfig` when combining the config file with what is passed inline to the `build`/`dev` command. The result is used for resolving.
- It merges `AstroConfig` when calling `updateConfig` from an integration.

**Comment by SalahAdDin** [comment]
What's about testing library Astro?

**Reply by Fryuni** [reply]
We have exposed the two functions needed to implement the testing library back on v5.4, so now `@inox-tools/astro-tests` run entirely on stable Astro APIs and so can any other testing library. There hasn't been any new discussion regarding providing a first-party helper package like that as it din't get that much interest, which could be sent to that third-party library. (well third-party from Astro's perspective, first-party to me since I'm the maintainer)

**Reply by SalahAdDin** [reply]
Does it work with Astro 6?

**Reply by Fryuni** [reply]
The code works but your package manager will refuse it. I wasn't planning on cutting a release compatible with Astro v6 while it is in beta.

**Reply by SalahAdDin** [reply]
Wanna check this too...It would be great to have examples in documentation.
